File: bsp_engineering/bsp_engineering/report/production_requirement_report/production_requirement_report.js
// Copyright (c) 2026, Pixfar and contributors
// For license information, please see license.txt

frappe.query_reports["Production Requirement Report"] = {
	"filters": [
		{
			"fieldname": "item_group",
			"label": __("Item Group"),
			"fieldtype": "MultiSelectList",
			"options": "Item Group",
			"get_data": function (txt) {
				return frappe.db.get_link_options("Item Group", txt);
			},
		},
		{
			"fieldname": "item_code",
			"label": __("Item"),
			"fieldtype": "MultiSelectList",
			"options": "Item",
			"get_data": function (txt) {
				return frappe.db.get_link_options("Item", txt);
			},
		},
		{
			"fieldname": "production_group",
			"label": __("Production Group"),
			"fieldtype": "MultiSelectList",
			"options": "Production Group",
			"get_data": function (txt) {
				return frappe.db.get_link_options("Production Group", txt);
			},
		},
		{
			"fieldname": "stock_status",
			"label": __("Stock Status"),
			"fieldtype": "Select",
			"options": [
				{ "value": "", "label": __("All") },
				{ "value": "Low Stock", "label": __("Low Stock (Red)") },
				{ "value": "Sufficient Stock", "label": __("Sufficient Stock") },
			],
		},
	],

	"onload": function (report) {
		// 1. Handle Column Persistence Logic
		frappe.call({
			method: "bsp_engineering.bsp_engineering.api.report_settings.get_user_report_settings",
			args: { report_name: report.report_name },
			callback: function (r) {
				if (r.message) {
					const { hidden_columns = [], column_order = [] } = r.message;

					const filter_and_sort = () => {
						if (!report.columns) return;

						// Filter out hidden columns
						report.columns = report.columns.filter(col => !hidden_columns.includes(col.id));

						// Reorder columns based on saved order
						if (column_order.length) {
							const ordered = [];
							const remaining = [];

							report.columns.forEach(col => {
								const idx = column_order.indexOf(col.id);
								if (idx !== -1) {
									ordered.push({ ...col, sort_idx: idx });
								} else {
									remaining.push(col);
								}
							});

							ordered.sort((a, b) => a.sort_idx - b.sort_idx);
							report.columns = [...ordered.map(c => { delete c.sort_idx; return c; }), ...remaining];
						}
					};

					filter_and_sort();

					const original_refresh = report.refresh;
					if (original_refresh) {
						report.refresh = function (...args) {
							filter_and_sort();
							return original_refresh.apply(this, args);
						};
					}
				}
			}
		});

		// 2. Define the Pick Columns action to be reused
		report.open_pick_columns_dialog = function () {
			frappe.call({
				method: "bsp_engineering.bsp_engineering.api.report_settings.get_user_report_settings",
				args: { report_name: report.report_name },
				callback: function (r) {
					const current_settings = r.message || {};
					const hidden_columns = current_settings.hidden_columns || [];

					const d = new frappe.ui.Dialog({
						title: __("Pick Columns"),
						fields: [
							{
								label: __("Select Columns"),
								fieldname: "columns",
								fieldtype: "MultiCheck",
								options: report.columns.map(col => ({
									label: col.label,
									value: col.id,
									checked: !hidden_columns.includes(col.id)
								}))
							}
						],
						primary_action: (values) => {
							const selected = values.columns || [];
							const new_hidden = report.columns
								.filter(col => !selected.includes(col.id))
								.map(col => col.id);

							const still_hidden = hidden_columns.filter(id => !report.columns.find(c => c.id === id));

							frappe.call({
								method: "bsp_engineering.bsp_engineering.api.report_settings.save_user_report_settings",
								args: {
									report_name: report.report_name,
									hidden_columns: [...new_hidden, ...still_hidden],
									column_order: []
								},
								callback: () => {
									report.refresh();
									d.hide();
								}
							});
						}
					});
					d.show();
				}
			});
		};

		// 3. Add to Report Menu (The "..." menu)
		if (report.get_menu_items) {
			const original_get_menu = report.get_menu_items;
			report.get_menu_items = function () {
				const items = original_get_menu ? original_get_menu.call(this) : [];
				items.push({
					label: __("Pick Columns"),
					action: () => report.open_pick_columns_dialog()
				});
				return items;
			};
		}

		// 4. Direct DOM Injection using MutationObserver (The most robust way for the toolbar)
		const observer = new MutationObserver(() => {
			const toolbar = $(".page-head-actions");
			if (toolbar.length && !$("#btn-pick-columns").length) {
				const btn = $(`<button class="btn btn-xs btn-default btn-pick-columns" style="margin-right: 10px; background-color: #fff; color: #000; border: 1px solid #d1d88d;">${__("Pick Columns")}</button>`);

				btn.on("click", () => {
					report.open_pick_columns_dialog();
				});
				toolbar.prepend(btn);
			}
		});

		observer.observe(document.body, {
			childList: true,
			subtree: true
		});
	},

	"formatter": function (value, row, column, data, default_formatter) {
		if (column.fieldname === "item_code" && data && data.item_code) {
			const code = frappe.utils.escape_html(data.item_code);
			return `<a href="/app/item/${encodeURIComponent(data.item_code)}" data-doctype="Item" data-name="${code}">${code}</a>`;
		}
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "total_stock" && flt(data.total_stock) < flt(data.total_low_qty)) {
			value = `<span style="color: var(--red-500, #d1242f); font-weight: 700;">${value}</span>`;
		}
		return value;
	},

	"after_datatable_render": function (report) {
		const datatable = report.datatable;
		if (!datatable) return;

		const original_remove = datatable.events.onRemoveColumn;
		datatable.events.onRemoveColumn = function (column) {
			if (original_remove) original_remove.call(this, column);

			frappe.call({
				method: "bsp_engineering.bsp_engineering.api.report_settings.get_user_report_settings",
				args: { report_name: report.report_name },
				callback: function (r) {
					const settings = r.message || {};
					const hidden = settings.hidden_columns || [];
					if (!hidden.includes(column.id)) {
						hidden.push(column.id);
					}

					frappe.call({
						method: "bsp_engineering.bsp_engineering.api.report_settings.save_user_report_settings",
						args: {
							report_name: report.report_name,
							hidden_columns: hidden,
							column_order: datatable.datamanager.getColumns(true).map(col => col.id)
						}
					});
				}
			});
		};

		const original_switch = datatable.events.onSwitchColumn;
		datatable.events.onSwitchColumn = function (col1, col2) {
			if (original_switch) original_switch.call(this, col1, col2);

			frappe.call({
				method: "bsp_engineering.bsp_engineering.api.report_settings.get_user_report_settings",
				args: { report_name: report.report_name },
				callback: function (r) {
					const settings = r.message || {};
					const current_order = datatable.datamanager.getColumns(true).map(col => col.id);

					frappe.call({
						method: "bsp_engineering.bsp_engineering.api.report_settings.save_user_report_settings",
						args: {
							report_name: report.report_name,
							hidden_columns: settings.hidden_columns || [],
							column_order: current_order
						}
					});
				}
			});
		};
	},
};

File: bsp_engineering/bsp_engineering/report/production_requirement_report/production_requirement_report.py
# Copyright (c) 2026, Pixfar and contributors
# For license information, please see license.txt

"""Production Requirement Report -- a fixed-warehouse copy of Low Stock and
Stock Summary Report ("For Dalai Order"). Same Low Qty/Stock/Total columns
and same red "needs a Dalai order" flag, but instead of a warehouse picker
it always shows exactly these three warehouses, in this fixed order, since
this report exists specifically to look at production-feeder stock:

    Noakhali Karkhana - BSP, Konapara Service Center - BSP, Store Room - BSP

See low_stock_and_stock_summary_report.py for the full column/data shape
this mirrors.

NOTE: the Report doctype's own "Add Total Row" checkbox must stay OFF for
this report, for the same reason as the report it's copied from -- see that
file's note.
"""

import frappe
from frappe import _
from frappe.query_builder import DocType
from frappe.utils import flt

from bsp_engineering.utils.item_category_sort import get_item_sort_map, sort_rows_by_item_category

# Fixed, in display order -- not user-selectable, unlike the report this is
# copied from.
FIXED_WAREHOUSES = [
	"Noakhali Karkhana - BSP",
	"Konapara Service Center - BSP",
	"Store Room - BSP",
]


def execute(filters=None):
	filters = frappe._dict(filters or {})
	warehouses = get_warehouses()
	columns = get_columns(warehouses)
	data = get_data(filters, warehouses)
	return columns, data


def get_warehouses():
	"""The fixed three warehouses, in FIXED_WAREHOUSES order -- looked up
	rather than hardcoding warehouse_name too, so a Warehouse rename is
	picked up automatically. A warehouse that's been deleted/renamed out
	from under FIXED_WAREHOUSES is silently skipped rather than erroring,
	so the report still renders with whichever of the three still exist."""
	wh_dt = DocType("Warehouse")
	rows = (
		frappe.qb.from_(wh_dt)
		.select(wh_dt.name, wh_dt.warehouse_name)
		.where(wh_dt.name.isin(FIXED_WAREHOUSES))
		.run(as_dict=True)
	)
	by_name = {row.name: row for row in rows}
	return [by_name[name] for name in FIXED_WAREHOUSES if name in by_name]


def get_columns(warehouses):
	columns = [
		{"label": _("Item Code"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 150},
		{"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Data", "width": 200},
		{
			"label": _("Item Group"),
			"fieldname": "item_group",
			"fieldtype": "Link",
			"options": "Item Group",
			"width": 130,
		},
	]

	for idx, wh in enumerate(warehouses):
		label = wh.warehouse_name or wh.name
		columns.append(
			{
				"label": _("{0} Low Qty").format(label),
				"fieldname": f"wh_{idx}_low_qty",
				"fieldtype": "Float",
				"width": 130,
			}
		)
		columns.append(
			{
				"label": _("{0} Stock").format(label),
				"fieldname": f"wh_{idx}_stock",
				"fieldtype": "Float",
				"width": 130,
			}
		)

	columns.append(
		{"label": _("Total Low Quantity"), "fieldname": "total_low_qty", "fieldtype": "Float", "width": 130}
	)
	columns.append({"label": _("Total Stock"), "fieldname": "total_stock", "fieldtype": "Float", "width": 130})
	return columns


def get_items(filters, warehouse_names):
	"""Every item in the system (subject only to the Item Group/Item/
	Production Group filters below) -- not just ones with a configured Item
	Low Stock Alert row. Previously scoped to that alert-configured universe
	(226 of 1094 items) the same way Low Stock Alert Report itself draws
	from; reported live as "not getting all the item reports" and changed
	by explicit request to show every item, no exceptions -- an item with
	no alert configured for these three warehouses just shows a real Stock
	figure from Bin (see get_data below) alongside a Low Qty of 0, rather
	than being left out of the report entirely."""
	if not warehouse_names:
		return []

	item_dt = DocType("Item")

	query = (
		frappe.qb.from_(item_dt)
		.select(item_dt.name, item_dt.item_name, item_dt.item_group)
		.orderby(item_dt.item_name)
	)

	item_groups = filters.get("item_group")
	if item_groups:
		if isinstance(item_groups, str):
			item_groups = frappe.parse_json(item_groups)
		if item_groups:
			query = query.where(item_dt.item_group.isin(item_groups))

	item_codes = filters.get("item_code")
	if item_codes:
		if isinstance(item_codes, str):
			item_codes = frappe.parse_json(item_codes)
		if item_codes:
			query = query.where(item_dt.name.isin(item_codes))

	production_groups = filters.get("production_group")
	if production_groups:
		if isinstance(production_groups, str):
			production_groups = frappe.parse_json(production_groups)
		if production_groups:
			query = query.where(item_dt.custom_production_group.isin(production_groups))

	return query.run(as_dict=True)


def get_data(filters, warehouses):
	warehouse_names = [wh.name for wh in warehouses]
	items = get_items(filters, warehouse_names)
	if not items:
		return []

	item_codes = [item.name for item in items]

	alert_dt = DocType("Item Low Stock Alert")
	low_qty_by_item_wh = {}
	for row in (
		frappe.qb.from_(alert_dt)
		.select(alert_dt.parent, alert_dt.warehouse, alert_dt.low_stock_qty)
		.where(alert_dt.parenttype == "Item")
		.where(alert_dt.parent.isin(item_codes))
		.where(alert_dt.warehouse.isin(warehouse_names))
		.run(as_dict=True)
	):
		low_qty_by_item_wh[(row.parent, row.warehouse)] = flt(row.low_stock_qty)

	bin_dt = DocType("Bin")
	stock_by_item_wh = {}
	for row in (
		frappe.qb.from_(bin_dt)
		.select(bin_dt.item_code, bin_dt.warehouse, bin_dt.actual_qty)
		.where(bin_dt.item_code.isin(item_codes))
		.where(bin_dt.warehouse.isin(warehouse_names))
		.run(as_dict=True)
	):
		stock_by_item_wh[(row.item_code, row.warehouse)] = flt(row.actual_qty)

	data = []
	for item in items:
		row = {"item_code": item.name, "item_name": item.item_name, "item_group": item.item_group}
		total_low_qty = 0.0
		total_stock = 0.0
		for wh_idx, wh in enumerate(warehouses):
			low_qty = low_qty_by_item_wh.get((item.name, wh.name), 0.0)
			stock = stock_by_item_wh.get((item.name, wh.name), 0.0)
			row[f"wh_{wh_idx}_low_qty"] = low_qty
			row[f"wh_{wh_idx}_stock"] = stock
			total_low_qty += low_qty
			total_stock += stock
		row["total_low_qty"] = total_low_qty
		row["total_stock"] = total_stock
		data.append(row)

	# "Filter using color" -- restrict to the same red/not-red split the
	# formatter (production_requirement_report.js) highlights with, rather
	# than just filtering visually. Applied last since it depends on the
	# totals just computed above.
	stock_status = filters.get("stock_status")
	if stock_status == "Low Stock":
		data = [row for row in data if flt(row["total_stock"]) < flt(row["total_low_qty"])]
	elif stock_status == "Sufficient Stock":
		data = [row for row in data if flt(row["total_stock"]) >= flt(row["total_low_qty"])]

	# BSP's official item-category order (see item_category_sort.py), replacing
	# the item_name ordering get_items() queried in.
	data = sort_rows_by_item_category(data, get_item_sort_map())
	return data
